File: scripts/validate_medistack_v0_1_export.py
#!/usr/bin/env python3
"""
validate_medistack_v0_1_export.py
MediStack v0.1 beta export contract validator.

목적: medistack_v0.1_beta_export.json 이 앱 배포 전에 v0.1 봉인 조건을
위반하지 않는지 자동 검출한다. 외부 의존성 없음(표준 라이브러리만).

사용:
    python3 validate_medistack_v0_1_export.py [path/to/medistack_v0.1_beta_export.json]

종료 코드: 0 = PASS, 1 = FAIL  (CI/배포 전 훅에서 사용)
"""
import sys
import json
import re

DEFAULT_PATH = "medistack_v0.1_beta_export.json"

EXPECTED_TOP_KEYS = {"meta", "disclaimers", "relations", "excluded_v0_1"}
EXPECTED_RELATIONS = 19
EXCLUDED_ID = 15
EXCLUDED_INGREDIENT = "에스오메프라졸"
EXCLUDED_NUTRIENT = "비타민B12"
POTASSIUM_NUTRIENT = "칼륨"
POTASSIUM_IDS = {17, 19}
FORBIDDEN_RELATION_FIELDS = ("status", "published", "clinical_reviewed")

# 제품 링크/예시 의심 필드(heuristic). product_link_allowed(플래그)는 제외.
# 스키마 확장 시 이 정규식/예외목록만 손보면 됨.
PRODUCT_FIELD_HINT = re.compile(r"(product|affiliate|shop|buy)", re.IGNORECASE)
PRODUCT_FIELD_SAFE = {"product_link_allowed"}


class Validator:
    def __init__(self):
        self.failures = []  # (no, title, detail)
        self.passes = []    # (no, title)

    def check(self, ok, no, title, detail=""):
        if ok:
            self.passes.append((no, title))
        else:
            self.failures.append((no, title, detail))
        return ok


def rid(obj):
    """relation/excluded 객체에서 id 또는 row_id 추출. 없으면 None."""
    if not isinstance(obj, dict):
        return None
    if "id" in obj:
        return obj["id"]
    if "row_id" in obj:
        return obj["row_id"]
    return None


def nonempty(v):
    if v is None:
        return False
    if isinstance(v, str):
        return v.strip() != ""
    if isinstance(v, (list, dict, tuple, set)):
        return len(v) > 0
    return True  # 숫자/불리언은 '존재'로 간주


def main(path):
    v = Validator()

    # 12) JSON 파싱 (실패 시 즉시 FATAL FAIL)
    try:
        with open(path, "r", encoding="utf-8") as f:
            data = json.load(f)
    except FileNotFoundError:
        print(f"[FATAL] 파일 없음: {path}")
        return 1
    except json.JSONDecodeError as e:
        print(f"[FATAL] JSON 파싱 실패: {e}")
        return 1

    if not isinstance(data, dict):
        print("[FATAL] 최상위가 객체(dict)가 아님")
        return 1

    # 1) top-level keys
    keys = set(data.keys())
    missing = EXPECTED_TOP_KEYS - keys
    extra = keys - EXPECTED_TOP_KEYS
    v.check(not missing and not extra, 1, "top-level keys (meta/disclaimers/relations/excluded_v0_1)",
            f"missing={sorted(missing)} extra={sorted(extra)}" if (missing or extra) else "")

    meta = data.get("meta")
    disc = data.get("disclaimers")
    rels = data.get("relations")
    excl = data.get("excluded_v0_1")

    # 12) 타입 가드
    meta_ok = isinstance(meta, dict)
    disc_ok = isinstance(disc, dict)
    rels_ok = isinstance(rels, list)
    excl_ok = isinstance(excl, list)
    type_problems = []
    if not meta_ok: type_problems.append("meta!=dict")
    if not disc_ok: type_problems.append("disclaimers!=dict")
    if not rels_ok: type_problems.append("relations!=list")
    if not excl_ok: type_problems.append("excluded_v0_1!=list")
    v.check(not type_problems, 12, "JSON 파싱/타입", "; ".join(type_problems) if type_problems else "")

    # 2) relations length = 19
    if rels_ok:
        v.check(len(rels) == EXPECTED_RELATIONS, 2, "relations length = 19", f"실제 {len(rels)}건")

    # 3) excluded_v0_1 = 15행 1건
    if excl_ok:
        e15 = [e for e in excl if rid(e) == EXCLUDED_ID]
        ok3 = (len(excl) == 1 and len(e15) == 1 and isinstance(e15[0], dict)
               and e15[0].get("ingredient") == EXCLUDED_INGREDIENT
               and e15[0].get("nutrient") == EXCLUDED_NUTRIENT)
        v.check(ok3, 3, "excluded_v0_1 = 15행(에스오메프라졸×비타민B12) 1건",
                f"excluded {len(excl)}건 / 15행 매칭 {len(e15)}건" if not ok3 else "")

    # 4) relations에 row_id 15 없음
    if rels_ok:
        leaked = [rid(r) for r in rels if rid(r) == EXCLUDED_ID]
        v.check(not leaked, 4, "relations에 row_id 15 미포함", f"15행 누출 {len(leaked)}건" if leaked else "")

    # 5) meta.published === false
    if meta_ok:
        pub = meta.get("published", "MISSING")
        v.check(pub is False, 5, "meta.published === false", f"값={pub!r}")
        # 6) meta.clinical_reviewed === false
        cr = meta.get("clinical_reviewed", "MISSING")
        v.check(cr is False, 6, "meta.clinical_reviewed === false", f"값={cr!r}")

    # 7) relation에 status/published/clinical_reviewed 필드 없음
    if rels_ok:
        offenders = []
        for r in rels:
            if isinstance(r, dict):
                bad = [k for k in FORBIDDEN_RELATION_FIELDS if k in r]
                if bad:
                    offenders.append((rid(r), bad))
        v.check(not offenders, 7, "relation에 status/published/clinical_reviewed 필드 없음",
                f"위반 {offenders}" if offenders else "")

    # 8) disclaimers.common 존재
    if disc_ok:
        v.check(nonempty(disc.get("common")), 8, "disclaimers.common 존재",
                "비어있거나 없음" if not nonempty(disc.get("common")) else "")
        # 9) disclaimers.potassium_notice 존재
        v.check(nonempty(disc.get("potassium_notice")), 9, "disclaimers.potassium_notice 존재",
                "비어있거나 없음" if not nonempty(disc.get("potassium_notice")) else "")

    # 10) 칼륨 17·19: product_link_allowed=false & potassium_safety_card=true
    if rels_ok:
        k_rels = [r for r in rels if isinstance(r, dict) and r.get("nutrient") == POTASSIUM_NUTRIENT]
        k_ids = {rid(r) for r in k_rels}
        problems = []
        if k_ids != POTASSIUM_IDS:
            problems.append(f"칼륨 id집합={sorted(k_ids)} (기대 {sorted(POTASSIUM_IDS)})")
        for r in k_rels:
            if r.get("product_link_allowed") is not False:
                problems.append(f"id{rid(r)} product_link_allowed={r.get('product_link_allowed')!r}")
            if r.get("potassium_safety_card") is not True:
                problems.append(f"id{rid(r)} potassium_safety_card={r.get('potassium_safety_card')!r}")
        v.check(not problems, 10, "칼륨 17·19 product_link_allowed=false & potassium_safety_card=true",
                "; ".join(problems) if problems else "")

    # 11) product_link_allowed=false 행에 제품링크/예시 필드 없음(+ 칼륨카드 일관성)
    if rels_ok:
        viol = []
        for r in rels:
            if isinstance(r, dict) and r.get("product_link_allowed") is False:
                for k, val in r.items():
                    if k in PRODUCT_FIELD_SAFE:
                        continue
                    if PRODUCT_FIELD_HINT.search(k) and nonempty(val):
                        viol.append(f"id{rid(r)} '{k}'={val!r}")
        inconsist = [rid(r) for r in rels if isinstance(r, dict)
                     and r.get("potassium_safety_card") is True
                     and r.get("product_link_allowed") is not False]
        if inconsist:
            viol.append(f"potassium_safety_card=true인데 link!=false: id{inconsist}")
        v.check(not viol, 11, "product_link_allowed=false 행에 제품링크/예시 필드 없음",
                "; ".join(viol) if viol else "")

    # ---- 결과 출력 ----
    total = len(v.passes) + len(v.failures)
    overall = "PASS" if not v.failures else "FAIL"
    print("=" * 62)
    print(f"MediStack v0.1 export 검증: {path}")
    print("=" * 62)
    if v.failures:
        print(f"\n[FAIL] {len(v.failures)}건")
        for no, title, detail in sorted(v.failures):
            print(f"  X #{no:<2} {title}" + (f"\n         -> {detail}" if detail else ""))
    else:
        print("\n모든 검증 통과.")
    print(f"\nRESULT: {overall}  ({len(v.passes)}/{total} checks passed)")
    print("=" * 62)
    return 0 if overall == "PASS" else 1


if __name__ == "__main__":
    p = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_PATH
    sys.exit(main(p))

File: scripts/validate_medistack_v0_1_export.js
#!/usr/bin/env node
/*
 * validate_medistack_v0_1_export.js
 * MediStack v0.1 beta export contract validator (Node, no deps).
 *
 * 사용:
 *   node validate_medistack_v0_1_export.js [path/to/medistack_v0.1_beta_export.json]
 *
 * 종료 코드: 0 = PASS, 1 = FAIL
 */
"use strict";
const fs = require("fs");

const DEFAULT_PATH = "medistack_v0.1_beta_export.json";
const EXPECTED_TOP_KEYS = ["meta", "disclaimers", "relations", "excluded_v0_1"];
const EXPECTED_RELATIONS = 19;
const EXCLUDED_ID = 15;
const EXCLUDED_INGREDIENT = "에스오메프라졸";
const EXCLUDED_NUTRIENT = "비타민B12";
const POTASSIUM_NUTRIENT = "칼륨";
const POTASSIUM_IDS = [17, 19];
const FORBIDDEN_RELATION_FIELDS = ["status", "published", "clinical_reviewed"];
const PRODUCT_FIELD_HINT = /(product|affiliate|shop|buy)/i;
const PRODUCT_FIELD_SAFE = new Set(["product_link_allowed"]);

const passes = [];
const failures = [];
function check(ok, no, title, detail = "") {
  if (ok) passes.push([no, title]);
  else failures.push([no, title, detail]);
  return ok;
}
function rid(o) {
  if (o === null || typeof o !== "object" || Array.isArray(o)) return null;
  if ("id" in o) return o.id;
  if ("row_id" in o) return o.row_id;
  return null;
}
function isPlainObj(o) {
  return o !== null && typeof o === "object" && !Array.isArray(o);
}
function nonempty(v) {
  if (v === null || v === undefined) return false;
  if (typeof v === "string") return v.trim() !== "";
  if (Array.isArray(v)) return v.length > 0;
  if (typeof v === "object") return Object.keys(v).length > 0;
  return true;
}
function setEq(a, b) {
  const A = new Set(a), B = new Set(b);
  if (A.size !== B.size) return false;
  for (const x of A) if (!B.has(x)) return false;
  return true;
}

function main(path) {
  let raw, data;
  try {
    raw = fs.readFileSync(path, "utf-8");
  } catch (e) {
    console.log(`[FATAL] 파일 없음/읽기 실패: ${path}`);
    return 1;
  }
  try {
    data = JSON.parse(raw);
  } catch (e) {
    console.log(`[FATAL] JSON 파싱 실패: ${e.message}`);
    return 1;
  }
  if (!isPlainObj(data)) {
    console.log("[FATAL] 최상위가 객체가 아님");
    return 1;
  }

  // 1) top-level keys
  const keys = Object.keys(data);
  const missing = EXPECTED_TOP_KEYS.filter((k) => !keys.includes(k));
  const extra = keys.filter((k) => !EXPECTED_TOP_KEYS.includes(k));
  check(missing.length === 0 && extra.length === 0, 1,
    "top-level keys (meta/disclaimers/relations/excluded_v0_1)",
    missing.length || extra.length ? `missing=[${missing}] extra=[${extra}]` : "");

  const meta = data.meta, disc = data.disclaimers, rels = data.relations, excl = data.excluded_v0_1;

  // 12) 타입 가드
  const metaOk = isPlainObj(meta), discOk = isPlainObj(disc),
        relsOk = Array.isArray(rels), exclOk = Array.isArray(excl);
  const tp = [];
  if (!metaOk) tp.push("meta!=object");
  if (!discOk) tp.push("disclaimers!=object");
  if (!relsOk) tp.push("relations!=array");
  if (!exclOk) tp.push("excluded_v0_1!=array");
  check(tp.length === 0, 12, "JSON 파싱/타입", tp.join("; "));

  // 2) relations length = 19
  if (relsOk) check(rels.length === EXPECTED_RELATIONS, 2, "relations length = 19", `실제 ${rels.length}건`);

  // 3) excluded = 15행 1건
  if (exclOk) {
    const e15 = excl.filter((e) => rid(e) === EXCLUDED_ID);
    const ok3 = excl.length === 1 && e15.length === 1 && isPlainObj(e15[0]) &&
      e15[0].ingredient === EXCLUDED_INGREDIENT && e15[0].nutrient === EXCLUDED_NUTRIENT;
    check(ok3, 3, "excluded_v0_1 = 15행(에스오메프라졸×비타민B12) 1건",
      ok3 ? "" : `excluded ${excl.length}건 / 15행 매칭 ${e15.length}건`);
  }

  // 4) relations에 15 없음
  if (relsOk) {
    const leaked = rels.filter((r) => rid(r) === EXCLUDED_ID);
    check(leaked.length === 0, 4, "relations에 row_id 15 미포함", leaked.length ? `15행 누출 ${leaked.length}건` : "");
  }

  // 5/6) meta 봉인
  if (metaOk) {
    check(meta.published === false, 5, "meta.published === false", `값=${JSON.stringify(meta.published)}`);
    check(meta.clinical_reviewed === false, 6, "meta.clinical_reviewed === false", `값=${JSON.stringify(meta.clinical_reviewed)}`);
  }

  // 7) relation 금지 필드 없음
  if (relsOk) {
    const offenders = [];
    for (const r of rels) {
      if (isPlainObj(r)) {
        const bad = FORBIDDEN_RELATION_FIELDS.filter((k) => k in r);
        if (bad.length) offenders.push(`id${rid(r)}:[${bad}]`);
      }
    }
    check(offenders.length === 0, 7, "relation에 status/published/clinical_reviewed 필드 없음",
      offenders.length ? `위반 ${offenders.join(", ")}` : "");
  }

  // 8/9) disclaimers
  if (discOk) {
    check(nonempty(disc.common), 8, "disclaimers.common 존재", nonempty(disc.common) ? "" : "비어있거나 없음");
    check(nonempty(disc.potassium_notice), 9, "disclaimers.potassium_notice 존재", nonempty(disc.potassium_notice) ? "" : "비어있거나 없음");
  }

  // 10) 칼륨 17·19
  if (relsOk) {
    const kRels = rels.filter((r) => isPlainObj(r) && r.nutrient === POTASSIUM_NUTRIENT);
    const kIds = kRels.map(rid);
    const problems = [];
    if (!setEq(kIds, POTASSIUM_IDS)) problems.push(`칼륨 id집합=[${kIds.sort()}] (기대 [${POTASSIUM_IDS}])`);
    for (const r of kRels) {
      if (r.product_link_allowed !== false) problems.push(`id${rid(r)} product_link_allowed=${JSON.stringify(r.product_link_allowed)}`);
      if (r.potassium_safety_card !== true) problems.push(`id${rid(r)} potassium_safety_card=${JSON.stringify(r.potassium_safety_card)}`);
    }
    check(problems.length === 0, 10, "칼륨 17·19 product_link_allowed=false & potassium_safety_card=true", problems.join("; "));
  }

  // 11) blocked 행 제품필드 없음 + 칼륨카드 일관성
  if (relsOk) {
    const viol = [];
    for (const r of rels) {
      if (isPlainObj(r) && r.product_link_allowed === false) {
        for (const k of Object.keys(r)) {
          if (PRODUCT_FIELD_SAFE.has(k)) continue;
          if (PRODUCT_FIELD_HINT.test(k) && nonempty(r[k])) viol.push(`id${rid(r)} '${k}'=${JSON.stringify(r[k])}`);
        }
      }
    }
    const inconsist = rels.filter((r) => isPlainObj(r) && r.potassium_safety_card === true && r.product_link_allowed !== false).map(rid);
    if (inconsist.length) viol.push(`potassium_safety_card=true인데 link!=false: id[${inconsist}]`);
    check(viol.length === 0, 11, "product_link_allowed=false 행에 제품링크/예시 필드 없음", viol.join("; "));
  }

  // ---- 출력 ----
  const total = passes.length + failures.length;
  const overall = failures.length === 0 ? "PASS" : "FAIL";
  const bar = "=".repeat(62);
  console.log(bar);
  console.log(`MediStack v0.1 export 검증: ${path}`);
  console.log(bar);
  if (failures.length) {
    console.log(`\n[FAIL] ${failures.length}건`);
    failures.sort((a, b) => a[0] - b[0]).forEach(([no, title, detail]) => {
      console.log(`  X #${String(no).padEnd(2)} ${title}` + (detail ? `\n         -> ${detail}` : ""));
    });
  } else {
    console.log("\n모든 검증 통과.");
  }
  console.log(`\nRESULT: ${overall}  (${passes.length}/${total} checks passed)`);
  console.log(bar);
  return overall === "PASS" ? 0 : 1;
}

process.exit(main(process.argv[2] || DEFAULT_PATH));
